lib/3scale_toolbox/openapi/oas3.rb

```diff
@@ -0,0 +1,232 @@
+module ThreeScaleToolbox
+  module OpenAPI
+    ##
+    #
+    # OAS3 object
+    # * OAS3.title -> string
+    # * OAS3.description -> string
+    # * OAS3.version -> string
+    # * OAS3.base_path -> string
+    # * OAS3.host -> string
+    # * OAS3.scheme -> string
+    # * OAS3.operation -> array of operation hash
+    #   * operation hash properties
+    #     * :verb
+    #     * :path
+    #     * :description
+    #     * :operation_id
+    # * OAS3.security ->  security hash
+    #   * security hash properties
+    #     * :id -> string
+    #     * :type -> string
+    #     * :name -> string
+    #     * :in_f -> string
+    #     * :flow -> symbol (:implicit_flow_enabled, :direct_access_grants_enabled, :service_accounts_enabled, :standard_flow_enabled)
+    #     * :scopes -> array of string
+    # * OAS3.service_backend_version -> string ('1','2','oidc')
+    # * OAS3.set_server_url -> def(spec, url)
+    # * OAS3.set_oauth2_urls-> def(spec, scheme_id, authorization_url, token_url)
+    class OAS3
+      META_SCHEMA_PATH = File.expand_path('../../../resources/oas3_meta_schema.json', __dir__)
+
+      def self.validate(raw)
+        meta_schema = JSON.parse(File.read(META_SCHEMA_PATH))
+        JSON::Validator.validate!(meta_schema, raw)
+      end
+
+      def self.build(path, raw, validate: true)
+        self.validate(raw) if validate
+
+        new(path, raw)
+      end
+
+      attr_reader :definition
+
+      def title
+        definition.info['title']
+      end
+
+      def description
+        definition.info['description']
+      end
+
+      def version
+        definition.info['version']
+      end
+
+      def base_path
+        # If there are many? take first
+        # From https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#openapi-object
+        # If the servers property is not provided, or is an empty array,
+        # the default value would be a Server Object with a url value of /
+        server_objects(&:path).first || '/'
+      end
+
+      def host
+        # If there are many? take first
+        server_objects(&:host).first
+      end
+
+      def scheme
+        # If there are many? take first
+        server_objects(&:scheme).first
+      end
+
+      def operations
+        @operations ||= parse_operations
+      end
+
+      def security
+        @security ||= parse_security
+      end
+
+      def service_backend_version
```

Comment:
Are we coupling 3scale specifics to the oas3/swagger classes? If that's intended then :+1: and will keep adding 3scale specific modifications/methods there

Reply:
Initially we had one class `ThreescaleOpenAPISpec` designed for that abstraction you mention. It did not serve very well. 

I decided to remove it and have oas3/swagger classes implement the same interface.  

```ruby
##                                                                                                                             
#                                                                                                                                   
# OpenAPI object                                                                                                                       
# * OpenAPI.title -> string                                                                                                            
# * OpenAPI.description -> string                                                                                                      
# * OpenAPI.version -> string                                                                                                          
# * OpenAPI.base_path -> string                                                                                                        
# * OpenAPI.host -> string                                                                                                             
# * OpenAPI.scheme -> string                                                                                                           
# * OpenAPI.operation -> array of operation hash                                                                                       
#   * operation hash properties                                                                                                     
#     * :verb                                                                                                                       
#     * :path                                                                                                                       
#     * :description                                                                                                                
#     * :operation_id                                                                                                               
# * OpenAPI.security ->  security hash                                                                                                 
#   * security hash properties                                                                                                      
#     * :id -> string                                                                                                               
#     * :type -> string                                                                                                             
#     * :name -> string                                                                                                             
#     * :in_f -> string                                                                                                             
#     * :flow -> symbol (:implicit_flow_enabled, :direct_access_grants_enabled, :service_accounts_enabled, :standard_flow_enabled)  
#     * :scopes -> array of string                                                                                                  
# * OpenAPI.service_backend_version -> string ('1','2','oidc')                                                                         
# * OpenAPI.set_server_url -> def(spec, url)                                                                                           
# * OpenAPI.set_oauth2_urls-> def(spec, scheme_id, authorization_url, token_url)                                                       
```

So my idea was to think about `ThreeScaleToolbox::OpenAPI::OAS3` and `ThreeScaleToolbox::OpenAPI::Swagger` not as OpenAPI parsers, but as 3scale openapi interface implementations.